EIPS/eip-2935.md

```diff
@@ -1,62 +1,122 @@
 
 ## Motivation
 
-There is increasingly a desire to remove the need for most clients to store history older than some relatively short duration (often between 1 week and 1 year) to save disk space. This requires some form of layer-2 network to help clients access historical information. These protocols can be made much simpler if blocks contained a quick Merkle path to historical blocks.
+Currently BLOCKHASH opcode accesses history to resolve hash of the block number in EVM. However a more stateless client friendly way is to maintain and serve these hashes from state.
 
-Additional secondary motivations include:
+Although this is possible even in Merkle trie state, but Verkle trie state further allows bundling the BLOCKHASH witnesses (along with other witnesses) in an efficient manner making it worthwhile to have these in state.
 
-* The protocol can be used to make more secure efficient light clients with flyclient-like technology (while the "optimal" flyclient protocol is fairly complex, large security gains over the status quo (trusted "canonical hash trees") can be made cheaply)
-* Improving cleanness of the protocol, as the BLOCKHASH opcode would then access state and not history.
+A side benefit of this approach could be that it allows building/validating proofs related to last 256 ancestors directly against the current state.
 
 ## Specification
 
 | Parameter | Value |
 | - | - |
-| `FORK_BLKNUM` | TBD |
+| `FORK_TIMESTAMP` | TBD |
 | `HISTORY_STORAGE_ADDRESS` | `0xfffffffffffffffffffffffffffffffffffffffe`|
+| `HISTORY_SERVE_WINDOW` | `256` |
 
-At the start of processing any block where `block.number > FORK_BLKNUM` (ie. before processing any transactions), run `sstore(HISTORY_STORAGE_ADDRESS, block.number - 1, block.prevhash)`.
+This EIP specifies for storing last `HISTORY_SERVE_WINDOW` block hashes in a ring buffer storage of `HISTORY_SERVE_WINDOW` length.
 
-When `block.number > FORK_BLKNUM + 256`, change the logic of the `BLOCKHASH` opcode as follows: if `FORK_BLKNUM <= arg < block.number`, return `sload(HISTORY_STORAGE_ADDRESS, arg)`. Otherwise return 0.
+
+At the start of processing any block where `block.timestamp >= FORK_TIMESTAMP` (ie. before processing any transactions), update the history in the following way:
+
+```python
+def process_block_hash_history(block: Block, state: State):
+    if block.timestamp >= FORK_TIMESTAMP:
+        state.insert_slot(HISTORY_STORAGE_ADDRESS, (block.number-1) % HISTORY_SERVE_WINDOW , block.parent.hash)
+
+    # If this is the first fork block, add the parent's direct 255 ancestors as well
+    if block.parent.timestamp < FORK_TIMESTAMP:
+        ancestor = block.parent
+        for i in range(HISTORY_SERVE_WINDOW - 1):
+            # stop at genesis block
+            if ancestor.number == 0:
+                break
+
+            ancestor = ancestor.parent
+            state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number % HISTORY_SERVE_WINDOW, ancestor.hash)
```

Comment:
this makes it look more analogous to L41

Reply:
actually, I think we should just rewrite it more clearly:

```python
for i in range(block.number-1, min(block.number-HISTORY_SERVE_WINDOW, 0)):
    state.insert_slot(HISTORY_STORAGE_ADDRESS, i % HISTORY_SERVE_WINDOW, ancestor.parent.hash)
    ancestor = ancestor.parent
```

then, the way it's implemented is left to the implementer.

Reply:
i think this is correct  storing at (`i-1`) than `i`, since 'i` is ancestor

or we shouldn't use i, just `ancestor.number -1` since that relationship is more clear (we always store ancestor's parent)

```python
    state.insert_slot(HISTORY_STORAGE_ADDRESS, (ancestor.number -1) % HISTORY_SERVE_WINDOW, ancestor.parent.hash)
```